File: Tools/workbench_oracles/distributions.py
#!/usr/bin/env python3
"""Oracle generator for the NumericSwift workbench — Distributions domain.

Computes bit-exact reference values with scipy.stats and freezes them as the
JSON fixture `Tests/NumericSwiftTests/Fixtures/workbench/distributions.json`.

Copied from the REFERENCE generator `integration.py`. Contract (WORKBENCH.md
§2/§3/§5):

  * ~100 cases partitioned 10 / 80 / 10 (trivial / hard / edge); the actual
    counts are printed so a thin tier is never silently shipped.
  * Each case carries `oracle.bits` (IEEE-754 hex) as the canonical value and a
    `source` citation. Oracle values come ONLY from scipy.stats — never from
    NumericSwift (FP1 / FP3 vacuous-gate rule).
  * The domain is single-strategy-per-function: each distribution function is a
    strategy id (`normal_cdf`, `t_ppf`, `chi2_cdf`, …). `inEnvelope` is per
    strategy. The one documented limitation (CLAUDE.md *Known Limitations §1*) is
    the Student-t `ppf` extreme tails (`p < 1e-4` or `p > 0.9999`): those cases
    are tagged `inEnvelope: {"t_ppf": false}`, so the gate requires NumericSwift
    to emit an `outsideEnvelope` diagnostic for them. Every other function/case
    is in-envelope (a correctness-vs-scipy gate).

Strategy ids and their parameters MUST match `registerDistributionsStrategies`
in `Sources/NumericSwiftWorkbenchKit/Domains/Distributions.swift`.

Run:
    /tmp/.nsoracle/bin/python Tools/workbench_oracles/distributions.py
"""

import json
import struct
import sys
from pathlib import Path

import scipy
from scipy import stats

SOURCE = f"scipy.stats {scipy.__version__}"


def bits_hex(value: float) -> str:
    """IEEE-754 bit pattern of a double as a 0x-prefixed 16-hex-digit string."""
    return "0x%016X" % struct.unpack("<Q", struct.pack("<d", value))[0]


# ── Oracle dispatch ─────────────────────────────────────────────────────────
# One scipy.stats frozen distribution per family. Each strategy id is
# "<family>_<func>"; the function is cdf / ppf / pdf.


def _frozen(family: str, params: dict):
    """Construct the scipy.stats frozen distribution for a family."""
    loc = params.get("loc", 0.0)
    scale = params.get("scale", 1.0)
    if family == "normal":
        return stats.norm(loc=loc, scale=scale)
    if family == "uniform":
        return stats.uniform(loc=loc, scale=scale)
    if family == "expon":
        return stats.expon(loc=loc, scale=scale)
    if family == "t":
        return stats.t(params["df"], loc=loc, scale=scale)
    if family == "chi2":
        return stats.chi2(params["df"], loc=loc, scale=scale)
    if family == "f":
        return stats.f(params["dfn"], params["dfd"], loc=loc, scale=scale)
    if family == "gamma":
        return stats.gamma(params["shape"], loc=loc, scale=scale)
    if family == "beta":
        return stats.beta(params["a"], params["b"], loc=loc, scale=scale)
    raise ValueError(f"unknown family {family}")


def oracle(family: str, func: str, arg: float, params: dict) -> float:
    """High-accuracy reference value from scipy.stats."""
    d = _frozen(family, params)
    return float(getattr(d, func)(arg))


def case(cid, tier, family, func, arg, *, params, tol, in_envelope=None):
    """Build one fixture entry.

    `arg` is the evaluation point: `x` for cdf/pdf, `p` for ppf. It is carried
    in the inputs bag under the function-appropriate key so the Swift strategy
    closure reads the same value.
    """
    strategy = f"{family_alias(family)}_{func}"
    inputs = {"func": func, "family": family_alias(family)}
    # cdf/pdf take x; ppf takes p.
    if func == "ppf":
        inputs["p"] = arg
    else:
        inputs["x"] = arg
    # Distribution parameters.
    for k, v in params.items():
        inputs[k] = v
    val = oracle(family, func, arg, params)
    # NaN/inf edge values: store bit-exact, write `value` as JSON null
    # (guard allow_nan=False at dump time like the other generators).
    json_value = val if (val == val and abs(val) != float("inf")) else None
    entry = {
        "id": cid,
        "tier": tier,
        "inputs": inputs,
        "oracle": {"value": json_value, "bits": bits_hex(val)},
        "source": SOURCE,
        "strategies": [strategy],
        "tol": {strategy: tol},
    }
    if in_envelope is not None:
        entry["inEnvelope"] = {strategy: in_envelope}
    return entry


# Swift-side strategy prefix per scipy family.
_ALIAS = {
    "normal": "normal",
    "uniform": "uniform",
    "expon": "expon",
    "t": "t",
    "chi2": "chi2",
    "f": "f",
    "gamma": "gamma",
    "beta": "beta",
}


def family_alias(family: str) -> str:
    return _ALIAS[family]


# Per-function default tolerances (absolute error vs scipy). cdf/pdf are
# closed-form-ish and very tight; ppf is Newton-Raphson and a touch looser.
TOL = {"cdf": 1e-10, "pdf": 1e-10, "ppf": 1e-7}


def build():
    cases = []

    # ── Trivial (~10): textbook smoke values, standard params ─────────────────
    trivial = [
        ("normal", "cdf", 0.0, {}),  # 0.5
        ("normal", "pdf", 0.0, {}),  # 1/sqrt(2pi)
        ("normal", "ppf", 0.5, {}),  # 0.0
        ("normal", "cdf", 1.96, {}),  # ~0.975
        ("uniform", "cdf", 0.5, {}),  # 0.5
        ("expon", "cdf", 1.0, {}),  # 1 - 1/e
        ("t", "cdf", 0.0, {"df": 10.0}),  # 0.5
        ("chi2", "cdf", 1.0, {"df": 1.0}),  # ~0.6827
        ("f", "cdf", 1.0, {"dfn": 5.0, "dfd": 10.0}),
        ("gamma", "cdf", 1.0, {"shape": 2.0}),
    ]
    for i, (fam, func, arg, params) in enumerate(trivial):
        cases.append(
            case(
                f"distributions.trivial.{family_alias(fam)}_{func}_{i}",
                "trivial",
                fam,
                func,
                arg,
                params=params,
                tol=TOL[func],
            )
        )

    # ── Hard (~80): realistic params / args across all families & funcs ───────
    # A grid of (family, func, [args], [param-sets]) yielding non-degenerate,
    # mid-distribution evaluations — the bulk of correctness coverage vs scipy.
    idx = 0
    # The grid is sized so the full sweep lands at ~80 cases with EVERY family
    # and EVERY function covered (no family is dropped by the cap). cdf carries
    # the most args (the bulk-correctness workhorse); pdf/ppf a few each.
    hard_specs = [
        # Normal — shifted/scaled.
        ("normal", "cdf", [-2.5, -1.0, 0.7, 2.3], [{"loc": 1.0, "scale": 2.0}]),
        ("normal", "pdf", [-1.0, 0.4, 1.8], [{"loc": 0.5, "scale": 1.5}]),
        ("normal", "ppf", [0.05, 0.25, 0.6, 0.9], [{"loc": 2.0, "scale": 3.0}]),
        # Student-t — central / near-tail (all in-envelope: |p| <= 0.9999).
        ("t", "cdf", [-3.0, -0.5, 1.2, 4.0], [{"df": 3.0}]),
        ("t", "pdf", [-2.0, 0.0, 2.5], [{"df": 5.0}, {"df": 100.0}]),
        ("t", "ppf", [0.01, 0.1, 0.4, 0.75, 0.95, 0.99], [{"df": 4.0}]),
        # Chi-squared.
        ("chi2", "cdf", [0.5, 2.0, 6.0, 15.0], [{"df": 3.0}]),
        ("chi2", "pdf", [1.0, 4.0, 9.0], [{"df": 2.0}]),
        ("chi2", "ppf", [0.05, 0.5, 0.9, 0.99], [{"df": 5.0}]),
        # F.
        ("f", "cdf", [0.5, 1.5, 3.0, 8.0], [{"dfn": 3.0, "dfd": 12.0}]),
        ("f", "pdf", [0.8, 2.0, 5.0], [{"dfn": 4.0, "dfd": 8.0}]),
        ("f", "ppf", [0.1, 0.5, 0.9, 0.99], [{"dfn": 5.0, "dfd": 15.0}]),
        # Gamma.
        ("gamma", "cdf", [0.5, 2.0, 5.0, 10.0], [{"shape": 2.0}]),
        ("gamma", "pdf", [1.0, 3.0, 7.0], [{"shape": 3.0}]),
        ("gamma", "ppf", [0.1, 0.5, 0.9], [{"shape": 2.5}]),
        # Beta.
        ("beta", "cdf", [0.2, 0.5, 0.8], [{"a": 2.0, "b": 5.0}, {"a": 0.5, "b": 0.5}]),
        ("beta", "pdf", [0.3, 0.6], [{"a": 3.0, "b": 2.0}]),
        ("beta", "ppf", [0.1, 0.5, 0.9], [{"a": 2.0, "b": 3.0}]),
        # Exponential / uniform tail of the grid.
        ("expon", "cdf", [0.5, 2.0, 5.0], [{"scale": 2.0}]),
        ("expon", "ppf", [0.1, 0.5, 0.9], [{"scale": 1.5}]),
        ("uniform", "cdf", [0.2, 0.5, 0.9], [{"loc": -1.0, "scale": 4.0}]),
        ("uniform", "ppf", [0.25, 0.75], [{"loc": 2.0, "scale": 6.0}]),
    ]
    for fam, func, args, param_sets in hard_specs:
        for params in param_sets:
            for arg in args:
                if idx >= 80:
                    break
                cases.append(
                    case(
                        f"distributions.hard.{family_alias(fam)}_{func}_{idx}",
                        "hard",
                        fam,
                        func,
                        arg,
                        params=params,
                        tol=TOL[func],
                    )
                )
                idx += 1

    # ── Edge (~10): the documented t_ppf extreme-tail out-of-envelope cases ───
    # CLAUDE.md Known Limitations §1: Student-t ppf is ~5 digits for |p| > 0.9999.
    # These MUST trip TDistribution.ppfDiagnosed's outsideEnvelope diagnostic.
    # Loose tol (1e-4) reflects the documented ~5-digit accuracy; the gate is
    # the *diagnostic*, not the numeric value.
    t_oov = [
        (1e-5, {"df": 5.0}),
        (5e-6, {"df": 10.0}),
        (1.0 - 1e-5, {"df": 5.0}),
        (1.0 - 5e-6, {"df": 20.0}),
        (1e-6, {"df": 3.0}),
    ]
    for i, (p, params) in enumerate(t_oov):
        cases.append(
            case(
                f"distributions.edge.t_ppf_oov_{i}",
                "edge",
                "t",
                "ppf",
                p,
                params=params,
                tol=1e-4,
                in_envelope=False,
            )
        )

    # In-envelope edge: t_ppf just INSIDE the envelope (|p| == 0.9999 boundary
    # region) and other-family extremes — must NOT emit a diagnostic, and must
    # stay within full-precision tol. This is the false-positive guard.
    edge_ok = [
        ("t", "ppf", 0.999, {"df": 5.0}, TOL["ppf"]),  # just inside (p < 0.9999)
        ("t", "ppf", 0.001, {"df": 5.0}, TOL["ppf"]),  # just inside (p > 1e-4)
        ("normal", "ppf", 0.99999, {}, 1e-7),  # normal has no documented tail limit
        ("chi2", "cdf", 50.0, {"df": 4.0}, 1e-10),  # far tail cdf, full precision
        ("gamma", "cdf", 30.0, {"shape": 2.0}, 1e-10),
    ]
    for i, (fam, func, arg, params, tol) in enumerate(edge_ok):
        cases.append(
            case(
                f"distributions.edge.{family_alias(fam)}_{func}_ok_{i}",
                "edge",
                fam,
                func,
                arg,
                params=params,
                tol=tol,
            )
        )

    return cases


def main():
    cases = build()
    tiers = {"trivial": 0, "hard": 0, "edge": 0}
    for c in cases:
        tiers[c["tier"]] += 1
    print(
        f"distributions: {len(cases)} cases — "
        f"trivial={tiers['trivial']} hard={tiers['hard']} edge={tiers['edge']}",
        file=sys.stderr,
    )

    out_dir = (
        Path(__file__).resolve().parents[2]
        / "Tests/NumericSwiftTests/Fixtures/workbench"
    )
    out_dir.mkdir(parents=True, exist_ok=True)
    out = out_dir / "distributions.json"
    # allow_nan=False: NaN/inf oracle values are stored bit-exact in `bits` and
    # written as JSON null in `value` (see `case`), so the dump must be finite.
    out.write_text(json.dumps(cases, indent=2, allow_nan=False) + "\n")
    print(f"wrote {out}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: Sources/NumericSwiftWorkbenchKit/Domains/Distributions.swift
//
//  Distributions.swift  (Domains/)
//  NumericSwiftWorkbenchKit
//
//  Per-domain workbench suite: Distributions (probability functions).
//
//  Mirrors the reference `Integration.swift` suite, but this is a
//  SINGLE-STRATEGY-PER-FUNCTION correctness domain (WORKBENCH.md §4): each
//  distribution function is its own strategy id — `normal_cdf`, `t_ppf`,
//  `chi2_pdf`, … — and the comparison scalar is that function's scalar output,
//  checked against scipy.stats.
//
//    1. `distributionsSuite` — the `DomainSuite` wiring the strategy + envelope
//       registries.
//    2. `registerDistributionsStrategies(into:)` — one closure per strategy id.
//       Each reads the fixture inputs bag (`family`, `func`, params, and the
//       evaluation point `x` or `p`), constructs the matching NumericSwift
//       distribution, and calls the requested function. Closures FORWARD any
//       ``NumericDiagnostic`` the library emitted — they never fabricate one.
//    3. `makeDistributionsEnvelopeRegistry()` — per-(strategy, tier) bounds.
//
//  ## Self-awareness
//
//  The one documented limitation (CLAUDE.md *Known Limitations §1*) is the
//  Student-t `ppf` extreme tails: only ~5 digits for `p < 1e-4` or `p > 0.9999`.
//  Those cases are tagged `inEnvelope: {"t_ppf": false}` in the fixture. The
//  `t_ppf` closure calls ``TDistribution/ppfDiagnosed(_:)`` (NOT the bare
//  ``TDistribution/ppf(_:)``) so the tail diagnostic propagates; the library —
//  not the harness — decides when to emit it. Every other function/case is
//  in-envelope (a correctness-vs-scipy gate), and t_ppf central cases must NOT
//  warn.
//
//  Inputs (carried through the JSON `inputs` bag, see ``InputValue``):
//    • `family` — distribution family tag (`normal`, `t`, `chi2`, `f`, `gamma`,
//                 `beta`, `expon`, `uniform`).
//    • `func`   — function name (`cdf`, `pdf`, `ppf`).
//    • `x` / `p`— scalar evaluation point (`x` for cdf/pdf, `p` for ppf).
//    • params   — distribution parameters: `df`, `dfn`, `dfd`, `shape`, `a`,
//                 `b`, `loc`, `scale` (each optional, family-dependent).
//
//  Licensed under the Apache License, Version 2.0.
//

import Foundation
import NumericSwift

extension Workbench {

    /// The Distributions domain suite.
    public static let distributionsSuite = DomainSuite(
        name: "distributions",
        registerStrategies: registerDistributionsStrategies,
        makeEnvelopeRegistry: makeDistributionsEnvelopeRegistry
    )
}

// MARK: - Input resolver

/// Common parameters pulled from a fixture inputs bag.
private struct DistInputs {
    let family: String
    let funcName: String
    let arg: Double          // x for cdf/pdf, p for ppf
    let loc: Double
    let scale: Double
    let df: Double?
    let dfn: Double?
    let dfd: Double?
    let shape: Double?
    let a: Double?
    let b: Double?
}

/// Extract the distribution inputs from a fixture bag, or `nil` when malformed.
@Sendable
private func distInputs(_ inputs: [String: InputValue]) -> DistInputs? {
    guard let family = inputs["family"]?.stringValue,
          let funcName = inputs["func"]?.stringValue
    else { return nil }
    // cdf/pdf carry `x`; ppf carries `p`.
    let arg: Double?
    if funcName == "ppf" {
        arg = inputs["p"]?.doubleValue
    } else {
        arg = inputs["x"]?.doubleValue
    }
    guard let argValue = arg else { return nil }
    return DistInputs(
        family: family,
        funcName: funcName,
        arg: argValue,
        loc: inputs["loc"]?.doubleValue ?? 0.0,
        scale: inputs["scale"]?.doubleValue ?? 1.0,
        df: inputs["df"]?.doubleValue,
        dfn: inputs["dfn"]?.doubleValue,
        dfd: inputs["dfd"]?.doubleValue,
        shape: inputs["shape"]?.doubleValue,
        a: inputs["a"]?.doubleValue,
        b: inputs["b"]?.doubleValue
    )
}

// MARK: - Per-family evaluation

/// Evaluate a non-t family's cdf/pdf/ppf, returning a bare scalar (no library
/// diagnostics defined for these — they are all in-envelope correctness cases).
@Sendable
private func evalNonT(_ d: DistInputs) -> Double? {
    switch d.family {
    case "normal":
        let dist = NormalDistribution(loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "uniform":
        let dist = UniformDistribution(loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "expon":
        let dist = ExponentialDistribution(loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "chi2":
        guard let df = d.df else { return nil }
        let dist = ChiSquaredDistribution(df: df, loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "f":
        guard let dfn = d.dfn, let dfd = d.dfd else { return nil }
        let dist = FDistribution(dfn: dfn, dfd: dfd, loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "gamma":
        guard let shape = d.shape else { return nil }
        let dist = GammaDistribution(shape: shape, loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    case "beta":
        guard let a = d.a, let b = d.b else { return nil }
        let dist = BetaDistribution(a: a, b: b, loc: d.loc, scale: d.scale)
        return scalarFor(d.funcName, dist.cdf, dist.pdf, dist.ppf, d.arg)
    default:
        return nil
    }
}

/// Pick the cdf/pdf/ppf member by name and apply it to `arg`.
@inline(__always)
private func scalarFor(
    _ funcName: String,
    _ cdf: (Double) -> Double,
    _ pdf: (Double) -> Double,
    _ ppf: (Double) -> Double,
    _ arg: Double
) -> Double? {
    switch funcName {
    case "cdf": return cdf(arg)
    case "pdf": return pdf(arg)
    case "ppf": return ppf(arg)
    default: return nil
    }
}

// MARK: - Strategy registrations

/// Populate `registry` with the Distributions strategies.
///
/// Strategy ids are `<family>_<func>`. The non-t families share `evalNonT`;
/// the t family splits ppf (the diagnosed path) from cdf/pdf.
@Sendable
public func registerDistributionsStrategies(into registry: inout StrategyRegistry) {

    // ── Non-t families: one closure per (family, func) — all bare scalars. ────
    let nonTFamilies = ["normal", "uniform", "expon", "chi2", "f", "gamma", "beta"]
    for family in nonTFamilies {
        for funcName in ["cdf", "pdf", "ppf"] {
            registry.register(id: "\(family)_\(funcName)") { inputs in
                guard let d = distInputs(inputs),
                      d.family == family, d.funcName == funcName,
                      let v = evalNonT(d)
                else { return nil }
                return StrategyResult(value: v)
            }
        }
    }

    // ── Student-t cdf/pdf: bare scalars (no documented envelope). ─────────────
    registry.register(id: "t_cdf") { inputs in
        guard let d = distInputs(inputs), d.funcName == "cdf", let df = d.df else { return nil }
        let dist = TDistribution(df: df, loc: d.loc, scale: d.scale)
        return StrategyResult(value: dist.cdf(d.arg))
    }
    registry.register(id: "t_pdf") { inputs in
        guard let d = distInputs(inputs), d.funcName == "pdf", let df = d.df else { return nil }
        let dist = TDistribution(df: df, loc: d.loc, scale: d.scale)
        return StrategyResult(value: dist.pdf(d.arg))
    }

    // ── Student-t ppf: the DIAGNOSED path — forwards the extreme-tail
    //    outsideEnvelope diagnostic the library emits for |p| > 0.9999. ────────
    registry.register(id: "t_ppf") { inputs in
        guard let d = distInputs(inputs), d.funcName == "ppf", let df = d.df else { return nil }
        let dist = TDistribution(df: df, loc: d.loc, scale: d.scale)
        let r = dist.ppfDiagnosed(d.arg)
        return StrategyResult(value: r.value, diagnostics: r.diagnostics)
    }
}

// MARK: - Envelope registry

/// Per-(strategy, tier) accuracy envelopes for the Distributions domain.
///
/// cdf/pdf are closed-form-ish (incomplete beta/gamma, Darwin transcendentals)
/// and reach near-machine precision; ppf is Newton-Raphson and a touch looser.
/// The fixture's per-case `tol` is authoritative (WORKBENCH.md §5); this
/// registry is the fallback when a case omits a tol for a strategy.
@Sendable
public func makeDistributionsEnvelopeRegistry() -> EnvelopeRegistry {
    var reg = EnvelopeRegistry()
    let families = ["normal", "uniform", "expon", "t", "chi2", "f", "gamma", "beta"]
    for tier: CaseTier in [.trivial, .hard, .edge] {
        for family in families {
            // cdf / pdf — tight everywhere.
            let cdfTol: Double = tier == .trivial ? 1e-12 : 1e-9
            reg.register(EnvelopeEntry(strategy: "\(family)_cdf", tier: tier, maxAbsError: cdfTol,
                description: "\(family) cdf vs scipy.stats — \(tier.rawValue) cases"))
            reg.register(EnvelopeEntry(strategy: "\(family)_pdf", tier: tier, maxAbsError: cdfTol,
                description: "\(family) pdf vs scipy.stats — \(tier.rawValue) cases"))

            // ppf — Newton-Raphson; full precision in-envelope, ~5 digits in the
            // documented t-ppf extreme tail (edge tier).
            let ppfTol: Double
            if family == "t", tier == .edge {
                ppfTol = 1e-4   // documented extreme-tail accuracy
            } else {
                ppfTol = tier == .trivial ? 1e-9 : 1e-6
            }
            reg.register(EnvelopeEntry(strategy: "\(family)_ppf", tier: tier, maxAbsError: ppfTol,
                description: "\(family) ppf vs scipy.stats — \(tier.rawValue) cases"))
        }
    }
    return reg
}
